### PR TITLE
[RW-6346][risk=low]: Link RAS account endpoint

### PR DESCRIPTION
**Pull request by yonghaoy**
Description:
This endpoint links RAS account with AoU account. We require to have a IAL2 login.gov account

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    @Bean
+    @Scope("prototype")
+    DbUser user() {
+      return currentUser;
+    }
+  }
+
+  @MockBean UserServiceAuditor mockUserServiceAuditor;
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:115]: Why did you need these? If it's for `UserService`, shouldn't those be documented here as required deps under `UserServiceTestConfiguration` - otherwise it's a bit misleading if that wasn't actually exhaustive (I'm not familiar with the class so I don't have the context).

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  @MockBean FireCloudService mockFireCloudService;
+  @MockBean ComplianceService mockComplianceService;
+  @MockBean DirectoryService mockDirectoryService;
+  @MockBean private FreeTierBillingService mockFreeTierBillingService;
+  @Mock private OpenIdConnectClient mockRasOidcClient;
+
+  @Before
+  public void setUp() throws Exception {
+    rasLinkService = new RasLinkService(userService, mockRasOidcClient);
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:124]: In an ideal test case, I think you would set up all the dependencies, and then `@Autowire` in your target service. To have that work in this case you would need to put `mockRasOidcClient` into the test configuration.

File: api/src/main/resources/workbench-api.yaml [hunk 3]
```diff
@@ -5161,12 +5161,16 @@ definitions:
...
+    - redirectUrl
     type: object
     properties:
       authCode:
         type: string
         description: 'the standard OAuth authorization code from RAS that we can use to exchange
         access token.'
+      redirectUrl:
+        type: string
+        description: 'The redirect url that was used when generating the authorization code.'
```
Comment (calbach) [inline api/src/main/resources/workbench-api.yaml:5173]: I'm not following this part - where are we getting this, how are we using it, and what is an example value in practice?

File: api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  public OpenIdConnectClient rasOidcClient(
+      CloudStorageClient cloudStorageClient, Provider<WorkbenchConfig> workbenchConfigProvider)
+      throws IOException {
+    String rasClientSecret = cloudStorageClient.getCredentialsBucketString(RAS_SECRET_BUCKET_NAME);
+    String rasClientId = workbenchConfigProvider.get().ras.clientId;
+    String rasTokenUrl = workbenchConfigProvider.get().ras.host + TOKEN_URL_SUFFIX;
+    String rasAuthorizeUrl = workbenchConfigProvider.get().ras.host + AUTHORIZE_URL_SUFFIX;
+    String rasUserInfoUrl = workbenchConfigProvider.get().ras.host + USER_INFO_URL_SUFFIX;
+
+    return new OpenIdConnectClient(
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:31]: Is this client safe for reuse, or concurrent use? If not, this should be `@RequestScope` (in general, this may be safer)

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  private static final TokenResponse TOKEN_RESPONSE_IAL2 =
+      new TokenResponse().setAccessToken(ACCESS_TOKEN).set(Id_TOKEN_FIELD_NAME, ID_TOKEN_JWT_IAL_2);
+
+  private long userId;
+  private static DbUser currentUser;
+
+  private RasLinkService rasLinkService;
+  private ObjectMapper objectMapper = new ObjectMapper();
+
+  @Autowired private UserService userService;
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:83]: A few things that make this test challenging:

1. UserService is a grab-bag of a bunch of random stuff, so it has a bunch of divergent dependencies.
2. you are instantiating a real implementation of UserService (not a mock). This makes for a better test, but also means you need to provide it with its dependencies. Personally, this also the approach I would probably also take.

I'm not sure any of this gets easier with another version of Spring boot. We could setup a partial config which mocks out all the deps of `UserService`, similar to `UserServiceTestConfiguration`, but I don't think it's a great idea to push mock definitions outside of the test itself. I think in an ideal world, there'd be a "testconfig" which worked without mocks, i.e. using fake in-memory implementations. This requires some work to build and scope out, though.

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    @Bean
+    @Scope("prototype")
+    DbUser user() {
+      return currentUser;
+    }
+  }
+
+  @MockBean UserServiceAuditor mockUserServiceAuditor;
+  @MockBean FireCloudService mockFireCloudService;
+  @MockBean ComplianceService mockComplianceService;
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:117]: Another way of delcaring these is on the test configuration, I don't remember whether this current approach you're using would allow injection into dependent services - the other approach may be required to make that work, but minimally this is more succinct syntax if you don't care about interacting with the mock at all. Example: https://github.com/all-of-us/workbench/blob/f56ff491ee1f303d434a89290e86dc944e6db0a4/api/src/test/java/org/pmiops/workbench/api/UserControllerTest.java#L72-L79

File: api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java (hunk 4 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:31]: Note: the default behavior for spring beans is **singleton**.

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java (hunk 1 above)
Comment (yonghaoy) [reply api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:115]: I am confused about this as well, and it took me a while to figure out what's going on. Those are all UserService dependencies, but all other tests like:
https://github.com/all-of-us/workbench/blob/master/api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java
https://github.com/all-of-us/workbench/blob/master/api/src/test/java/org/pmiops/workbench/api/RuntimeControllerTest.java
https://github.com/all-of-us/workbench/blob/master/api/src/test/java/org/pmiops/workbench/api/DataSetControllerTest.java
(basicly **all** tests that use UserServiceTestConfiguration are the same)
all have this pattern. I guess that is because UserServiceTestConfiguration hold them most common, simple dependencies, while each test will in charge of creating other beans, depends on whether they need mock or not? 
So I just follow other's pattern and follow the common pattern.

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java (hunk 2 above)
Comment (yonghaoy) [reply api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:124]: Done.
But another pattern that we use before:
is if the target service does not need bean, I just need @Autowire the target service.
But if we need to mock some service, or we want to change the target service's parameter. Then `@Autowire` or `@MockBean` all target's service parameters, then construct the target service on our own. This makes some tests easier and more flexible in some cases.

And also, I am not sure @TestConfiguration is a good candidate in the future, that also confused me.

File: api/src/main/resources/workbench-api.yaml (hunk 3 above)
Comment (yonghaoy) [reply api/src/main/resources/workbench-api.yaml:5173]: That is a tricky one, and I am not sure if I handled it correctly. 
(1) That is not used(the redirect URI happens in UI), but RAS requires it even for the following code exchange call (with &refirect_uri=), and that is the common pattern for OAuth. So I have to set it correctly to make RAS happy. 
(2) I could put it in config file, but I am not sure if that is better. It is relative easier for UI to get it, and just set it in the request. (in UI, it can grab the current url then append the static /ras-callback). So I just follow Bond's pattern https://github.com/DataBiosphere/bond/blob/develop/swagger/api-docs.yaml#L169. Terra UI grab the redirect uri, then Bond just consume it.

File: api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java (hunk 4 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:31]: I think this should be singleton, and better to be singleton. The main thing is `HttpTransport`, and I think that is a heavy-weight one. Their javadoc recommandation is: 
"For maximum efficiency, applications should use a single globally-shared instance of the HTTP transport.

Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:31]: btw, do you know why my api-local-test  are failing for some related `OpenIdConnectClient` scope error?

Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:31]: Nice, it is directly related to this discussion. The WorkbenchConfig value is request scoped (some discussion of that here: https://pmi-engteam.slack.com/archives/GP4CG4TFF/p1575930937326700?thread_ts=1575925918.323100&cid=GP4CG4TFF). The CloudStorageService depends on the config, therefore this injection will not work since your bean is singleton. Changing this to request scoped would be the easiest fix.

RE: HttpTransport, is there an alternate constructor that accepts that parameter? We already have a bean for that which you inject, and is already a singleton: https://github.com/all-of-us/workbench/blob/f56ff491ee1f303d434a89290e86dc944e6db0a4/api/src/main/java/org/pmiops/workbench/config/AppEngineConfig.java#L12

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java (hunk 5 above)
Comment (yonghaoy) [reply api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:83]: Now I feel I use UserService wrongly. I am looking for a layer between `RasLinkService` and `UserDao`, which handles the extra process when updating/reading from Db. What I really need is `updateUserWithRetries` with the function that also update other tables, also `Agent` stuff.  I can try to let RasService directly use `UserDao` with some extra logic, but I feel that is harder. 
What we can do is 
(1) split UserService into `UserStorageService` and `UserService`. 
(2) Further improve: pull those random stuff out into a small service. for example, like the RAS, put extra dependencies(OidcClietn) in its own service, and only use UserStorageService to do the final pure DB stuff. 

Step 1 might be a good start that we can discuss with team.

File: api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java (hunk 4 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:34]: Thanks ,changed to request scope.  HttpTransport, did a code search, seems most examples use that as a static final variable.

Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:34]: Injection seems preferred since we already even have a bean to do this. If we wanted to write a unit test for OpenIdConnectClient we would immediately need to deal with this static dependency

File: api/src/main/java/org/pmiops/workbench/ras/OpenIdConnectClient.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+public class OpenIdConnectClient {
+  private static final NetHttpTransport DEFAULT_HTTP_TRANSPORT = new NetHttpTransport();
+  private static final JacksonFactory DEFAULT_JACKSON_FACTORY = new JacksonFactory();
+
+  private final AuthorizationCodeFlow codeFlow;
+  private final String userInfoUrl;
+  private final ObjectMapper objectMapper;
+
+  OpenIdConnectClient(
+      String clientId,
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/OpenIdConnectClient.java:36]: This constructor signature seems very error-prone as a caller (5 strings in the exact correct order), can it be reduced at all? Aren't the token, authorize, and user info paths well known and could be constants in this class?

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    throw new ServerErrorException(
+        String.format("Invalid acl Claim in OIDC id token for %s", acrClaim));
+  }
+
+  /**
+   * Validates and extracts user's login.gov account from UserInfo response in Json format. See
+   * class javadoc Step3 for more details.
+   */
+  private static String getLoginGovUsername(JsonNode userInfo) {
+    String preferredUsername = userInfo.get(PREFERRED_USERNAME_FIELD_NAME).asText("");
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:136]: Mapping this into the expected structure, e.g. a POJO, could perhaps make this more readable

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+ *
+ * <pre>{@code
+ * "sub":"subId",
+ * "name" : "",
+ * "preferred_username":"user1@Login.Gov",
+ * "userId":"user1"
+ * "email":"foo@gmail.com"
+ * }</pre>
+ *
+ * The {@code preferred_username} field should end with "@login.gov" if using that to login. The
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:71]: What's the significance of the "preferred" username? Why do we check this? (please add to the comment)

(ideally the RAS sign in would just tell us all the accounts they have - we shouldn't care which one they happened to sign in with recently)

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+ * The acr claim contains user IAL status, the number after /assurance/ial/
+ *
+ * <p>Step3: User access_token to pull RAS user info. A sample userInfo in Json format:
+ *
+ * <pre>{@code
+ * "sub":"subId",
+ * "name" : "",
+ * "preferred_username":"user1@Login.Gov",
+ * "userId":"user1"
+ * "email":"foo@gmail.com"
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:68]: Is this just a random contact email value that we don't care about?

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+  private static final Logger log = Logger.getLogger(RasLinkService.class.getName());
+
+  private final UserService userService;
+  private final OpenIdConnectClient rasOidcClient;
+
+  @Autowired
+  public RasLinkService(
+      UserService userService, @Qualifier(RAS_OIDC_CLIENT) OpenIdConnectClient rasOidcClient) {
+    this.userService = userService;
+    this.rasOidcClient = rasOidcClient;
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:92]: If you want the service to be singleton, this client would need to be fetched lazily, via `Provider<OpenIdConnectClient>`, since it is request scoped. This is the usual pattern for services

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+      userInfoResponse = rasOidcClient.fetchUserInfo(tokenResponse.getAccessToken());
+    } catch (IOException e) {
+      log.log(Level.WARNING, "Failed to link RAS account", e);
+    }
+    return userService.updateRasLinkLoginGovStatus(getLoginGovUsername(userInfoResponse));
+  }
+
+  /** Validates user has IAL2 setup. See class javadoc Step2 for more details. */
+  static boolean isIal2(String acrClaim) {
+    Pattern p = Pattern.compile("ial/\\d", Pattern.CASE_INSENSITIVE);
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:122]: What is the format of the "acrClaim" string? The check here seems slightly specific - since it seems it would allow either a suffix or prefix string. I'd think we'd want to be looking for a very specific value here, and also understand how other values might be encoded in the future.

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+              .getClaim(ACR_CLAIM)
+              .asString();
+      if (!isIal2(acrClaim)) {
+        throw new ForbiddenException(
+            String.format("User does not have IAL2 enabled, acrClaim: %s", acrClaim));
+      }
+      // Fetch user info.
+      userInfoResponse = rasOidcClient.fetchUserInfo(tokenResponse.getAccessToken());
+    } catch (IOException e) {
+      log.log(Level.WARNING, "Failed to link RAS account", e);
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:115]: why do we  allow the method to continue to execute here?

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java (hunk 10 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:68]: that is the email we use to register and login in login.gov(serve as username). Based on my experience, that seems to be immutable, and unique per user. (you need to delete that account if you want to re-use that emai), and that is the only meaningful username we can get from login.gov. So I use that as login.gov username.
Another option is use user id, but I think that is just a randomUUID, that is not as useful as email.

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java (hunk 11 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:92]: RE: test failure, see this comment^

**Review by yonghaoy — COMMENTED** [review]
I think autowire the RasService in test does not change mock return value, tried reset(). 
I think it's okay to create a one with the mock we want. Feels that is a better option than DirectContext(After_each_method)?

File: api/src/main/java/org/pmiops/workbench/ras/OpenIdConnectClient.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+      return this;
+    }
+
+    public Builder setHttpTransport(HttpTransport httpTransport) {
+      this.httpTransport = httpTransport;
+      return this;
+    }
+
+    public OpenIdConnectClient build() throws IOException {
+      checkNotNull(clientId, String.format("Missing clientId"));
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/OpenIdConnectClient.java:149]: nit: string.format are unused throughout - can probably drop them

File: api/src/main/java/org/pmiops/workbench/ras/OpenIdConnectClient.java (hunk 7 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/ras/OpenIdConnectClient.java:37]: Yep, this is nice - thank you.

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkConstants.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+  // The login.gov's email field which we choose as login.gov username.
+  static final String EMAIl_FIELD_NAME = "email";
+
+  // The identifier that indicate users login using login.gov account.
+  static final String LOGIN_GOV_IDENTIFIER_LOWER_CASE = "@login.gov";
+
+  // The prefix in ACR claim that we can use to identify user IAL status.
+  // ACR_CLAIM_IAL_PREFIX + {1/2} is the user's IAL tier.
+  static final String ACR_CLAIM_IAL_PREFIX = "stsstg.nih.gov/assurance/ial/";
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkConstants.java:38]: Hm, do we actually care about the prefix? Might be simpler to just have a single string constant for the whole thing (ACR_CLAIM_IAL2), we don't really use IAL1 anyways

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+      throw new ServerErrorException("Failed to link RAS account", e);
+    }
+    return userService.updateRasLinkLoginGovStatus(getLoginGovUsername(userInfoResponse));
+  }
+
+  /** Validates user has IAL2 setup. See class javadoc Step2 for more details. */
+  static boolean isIal2(String acrClaim) {
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:125]: You said this is string delimited - then would it make sense to first break this string into individual tokens, and verify that at least one matches exactly?

It's unlikely to matter since presumably these ACR values are maintained by RAS, but it seems incorrect that the following claims are accepted by this regex check:
"myevilsite.com/https://stsstg.nih.gov/assurance/ial/2"
"https://stsstg.nih.gov/assurance/ial/27"

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+    Pattern p = Pattern.compile(ACR_CLAIM_IAL_PREFIX + "\\d", Pattern.CASE_INSENSITIVE);
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:126]: Is the spec actually case insensitive?

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java (hunk 8 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:140]: This is probably fine, if the responses get more complex, e.g. with passports, we should consider it though. Pulling keys out via string constants can be harder to read in comparison to a class with named properties - specifically there is not an easy place to look to see the full response structure (beyond comments, which can go out of sync).

File: api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Bean(RAS_OIDC_CLIENT)
+  @RequestScope(proxyMode = ScopedProxyMode.DEFAULT)
+  @Lazy
+  public OpenIdConnectClient rasOidcClient(
+      CloudStorageClient cloudStorageClient,
+      Provider<WorkbenchConfig> workbenchConfigProvider,
+      @Qualifier(RAS_HTTP_TRANSPORT) HttpTransport httpTransport)
+      throws IOException {
+    String rasClientSecret = cloudStorageClient.getCredentialsBucketString(RAS_SECRET_BUCKET_NAME);
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/ras/RasOidcClientConfig.java:34]: opt nit: I don't think the local vars are super helpful here - I might just use a `host` local var for readability, then inline everything in the builder below

File: api/src/main/resources/workbench-api.yaml (hunk 3 above)
Comment (calbach) [reply api/src/main/resources/workbench-api.yaml:5173]: What is an example value, for a request in the test env?

**Review by calbach — COMMENTED** [review]
I think the mostly looks good - just some more questions for my understanding

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java (hunk 17 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:126]: Seems I always get Login.Gov from RAS. But I am concern that is such a small piece in response and might be just changed behind the scene. So I use the lower case just in case it happens. Shall I make this case sensitive?

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java (hunk 16 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:125]: Changed to the exactly match. Thanks

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkConstants.java (hunk 15 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasLinkConstants.java:38]: I was thinking invalid acr vs IA1 different cases, and we want to treat them differently in UI. But I guess we just care 2 vs not 2.

File: api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java (hunk 8 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java:133]: The userInfo is a JWT with some common user information, like email, username, timestamp. Those are simple  enough that don't need pojo. I saw most Oidc library wrapper with those common field plus a generic get(String claimName);  I didn't implement that because the only common thing we need is '`email`'.
Speaking of Passport, that is a complicated Claim(Json), that we  probably want a dedicate pojo, with some method to understand passport. We can add when we need that.

**Review by yonghaoy — COMMENTED** [review]
Thanks!

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java [hunk 19]
```diff
@@ -613,7 +617,10 @@ private boolean userHasEverLoggedIn(
 
   @Override
   public ResponseEntity<Profile> linkRasAccount(RasLinkRequestBody body) {
-    // TODO(RW-6364): Implement this endpoint.
-    return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
+    DbUser dbUser =
+        rasLinkServiceProvider
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/api/ProfileController.java:621]: nit: since you are using a provider for your OIDC client, you don't need to use a provider here, it can just be a direct service

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+      return WorkbenchConfig.createEmptyConfig();
+    }
+
+    @Bean
+    @Scope("prototype")
+    DbUser user() {
+      return currentUser;
+    }
+  }
+
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:146]: If you use the pattern of:

```
@MockBean{
 ...
 OpenIdConnectClient.class
}
class TestConfiguration { 
...
}

@Autowired OpenIdConnectClient mockOidcClient;
```

That should function as a mock. The test config generates the mock, then you autowire (inject) it into your test class.

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    @Bean
+    HttpTransport httpTransport() {
+      return mockHttpTransport;
+    }
+
+    @Bean
+    @Qualifier(RAS_OIDC_CLIENT)
+    Provider<OpenIdConnectClient> rasOidcClientProvider() {
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:124]: You don't want to export a `Provider` as a bean, I don't think this does anything (or else, maybe it does something surprising). A `Provider` is a special class in Spring beans which can be used to dynamically inject a bean on-demand, rather than at instantiation time. You would just provide the specific type normally `OpenIdConnectClient`, and clients that call `oidcProvider.get()` will get that bean.

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+      return CLOCK;
+    }
+
+    @Bean
+    Random random() {
+      return new FakeLongRandom(123);
+    }
+
+    @Bean
+    HttpTransport httpTransport() {
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:118]: This should also not be needed, putting `HttpTransport.class` into the `MOckBean` array should suffice

**Review by calbach — APPROVED** [review]
Looks good, just a few test cleanups - let me know if you have more questions on that

File: api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java (hunk 20 above)
Comment (yonghaoy) [reply api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:136]: I tried that but I don't know how to put @Qualifier inside mockbean array. So I end up with doing the current way.

Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java:136]: Ahhh, OK, this makes sense then - I don't know off the top of my head a better pattern
